src/app/system/audit-trails/audit-trails.component.ts

```diff
@@ -0,0 +1,411 @@
+  /** Filtered action data for autocomplete. */
+  filteredActionData: any;
+  /** Filtered entity data for autocomplete. */
+  filteredEntityData: any;
+  /** Audit Trail Search Template Data. */
+  auditTrailSearchTemplateData: any;
+  /** Columns to be displayed in audit trails table. */
+  displayedColumns: string[] = ['id', 'resourceId', 'status', 'madeBy', 'actionName', 'entityName', 'officeName', 'madeOnDate', 'checker', 'checkedOnDate'];
+  /** Data source for audit trails table. */
+  dataSource: AuditTrailsDataSource;
+  /** Audit Trails filter. */
+  filterAuditTrailsBy = [
+    {
+      type: 'actionName',
+      value: ''
+    },
+    {
+      type: 'entityName',
+      value: ''
+    },
+    {
+      type: 'resourceId',
+      value: ''
+    },
+    {
+      type: 'makerId',
+      value: ''
+    },
+    {
+      type: 'makerDateTimeFrom',
+      value: ''
+    },
+    {
+      type: 'makerDateTimeTo',
+      value: ''
+    },
+    {
+      type: 'checkerDateTimeFrom',
+      value: ''
+    },
+    {
+      type: 'checkerDateTimeTo',
+      value: ''
+    },
+    {
+      type: 'checkerId',
+      value: ''
+    },
+    {
+      type: 'processingResult',
+      value: ''
+    },
+    {
+      type: 'dateFormat',
+      value: 'yyyy-MM-dd'
+    },
+    {
+      type: 'locale',
+      value: 'en'
+    }
+  ];
+  /** User form control. */
+  user = new FormControl('');
+  /** From date form control. */
+  fromDate = new FormControl();
+  /** Checked from date form control. */
+  checkedFromDate = new FormControl();
+  /** Processing result form control. */
+  processingResult = new FormControl();
+  /** Action name form control. */
+  actionName = new FormControl();
+  /** Resource ID form control. */
+  resourceId = new FormControl('');
+  /** To date form control. */
+  toDate = new FormControl();
+  /** Checked to date form control. */
+  checkedToDate = new FormControl();
+  /** Entity name form control. */
+  entityName = new FormControl();
+  /** Checker form control. */
+  checker = new FormControl();
+
+  /** Paginator for audit trails table. */
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+  /** Sorter for audit trails table. */
+  @ViewChild(MatSort) sort: MatSort;
+
+  /**
+   * Retrieves the audit trails and audit trail search template data from `resolve`.
+   * @param {ActivatedRoute} route Activated Route.
+   * @param {SystemService} systemService System Service.
+   */
+  constructor(private route: ActivatedRoute,
+              private systemService: SystemService) {
+    this.route.data.subscribe((data: { auditTrailSearchTemplate: any }) => {
+      this.auditTrailSearchTemplateData = data.auditTrailSearchTemplate;
+    });
+  }
+
+  /**
+   * Sets filtered users and actions for autocomplete and audit trails table.
+   */
+  ngOnInit() {
+    this.setFilteredUsers();
+    this.setFilteredActions();
+    this.setFilteredEntities();
+    this.getAuditTrails();
+    this.initAuditTrails();
```

Comment:
It would be better if we use global variable, otherwise I have to use many workarounds to get it running. I have made other required changes.

Reply:
We need to do this because it is taking time to update when the user clicks on `Download` button. It keeps the data ready for the user and the user doesn't have to wait. Please share your thoughts.

Reply:
No, we should do it only during download. We will be unnecessarily increasing overhead by sending multiple requests every time. It's alright for download to take a bit of time.